File: games/fill-in-the-blank/frontend/types/room.ts
export interface Player {
  id: string;
  display_name: string;
  avatar: string;
  is_host: boolean;
  is_ready: boolean;
  is_spectator: boolean;
  connected: boolean;
}

export interface Submission {
  submission_id: string;
  text: string;
}

export interface VoteResult {
  submission_id: string;
  text: string;
  author_id: string;
  votes: number;
}

export interface GameState {
  phase: string;
  question_number: number;
  total_questions: number;
  prompt: string | null;
  submitted_count: number;
  submissions: Submission[] | null;
  voted_count: number;
  results: VoteResult[] | null;
  scores: Record<string, number>;
}

export interface Room {
  code: string;
  game_type: string;
  is_private: boolean;
  phase: string;
  max_players: number;
  players: Player[];
  invite_url: string;
  game_state: GameState | null;
}

export interface RoomSummary {
  code: string;
  game_type: string;
  phase: string;
  is_private: boolean;
  player_count: number;
  max_players: number;
}

export interface CreateRoomResponse {
  room: Room;
  player_id: string;
}

File: games/fill-in-the-blank/backend/app/schemas/room.py
from pydantic import BaseModel


class CreateRoomRequest(BaseModel):
    game_type: str
    host_display_name: str
    host_avatar: str = "default"
    is_private: bool = False


class JoinRoomRequest(BaseModel):
    display_name: str
    avatar: str = "default"
    as_spectator: bool = False


class PlayerOut(BaseModel):
    id: str
    display_name: str
    avatar: str
    is_host: bool
    is_ready: bool
    is_spectator: bool
    connected: bool


class GameStateOut(BaseModel):
    """Where the active Fill in the Blank game currently is."""

    phase: str
    question_number: int = 0
    total_questions: int = 0
    prompt: str | None = None
    submitted_count: int = 0
    # [{submission_id, text}] populated after reveal — author_id intentionally absent until results
    submissions: list[dict] | None = None
    voted_count: int = 0
    # [{submission_id, text, author_id, votes}] populated in results phase only
    results: list[dict] | None = None
    scores: dict[str, int] = {}


class RoomOut(BaseModel):
    code: str
    game_type: str
    is_private: bool
    phase: str
    max_players: int
    players: list[PlayerOut]
    invite_url: str
    game_state: GameStateOut | None = None


class CreateRoomResponse(BaseModel):
    room: RoomOut
    player_id: str


class RoomSummary(BaseModel):
    code: str
    game_type: str
    phase: str
    is_private: bool
    player_count: int
    max_players: int
